File: LeetCode/0220. Contains Duplicate III/Solution.kt
import java.util.TreeSet

/**
 * 40 / 40 test cases passed.
 * Status: Accepted
 * Runtime: 418 ms
 */
class Solution {
    fun containsNearbyAlmostDuplicate(nums: IntArray, k: Int, t: Int): Boolean {
        val kRange = TreeSet<Long>()
        nums.forEachIndexed { index, num ->
            if (match(kRange, num, t)) return true

            kRange.add(num.toLong())

            if (kRange.size == k + 1) kRange.remove(nums[index - k].toLong())
        }
        return false
    }

    private fun match(nums: TreeSet<Long>, num: Int, t: Int): Boolean {
        return nums.ceiling(num.toLong() - t.toLong())?.let { it <= num.toLong() + t.toLong() } ?: false
    }
}

fun main(args: Array<String>) {
    Solution().containsNearbyAlmostDuplicate(intArrayOf(-1, -1), 1, 0).let(::println)
    Solution().containsNearbyAlmostDuplicate(intArrayOf(-2147483648,-2147483647), 3, 3).let(::println)
}

File: LeetCode/0220. Contains Duplicate III/solution.cpp
/**
 * 54 / 54 test cases passed.
 * Runtime: 68 ms
 * Memory Usage: 18.3 MB 
 */
class Solution {
public:
    bool containsNearbyAlmostDuplicate(vector<int>& nums, int k, int t) {
        set<long long> slide;
        for (int i = 0; i < nums.size(); ++ i) {
            /*
            判断滑动窗口中是否存在某个数 y 落在区间 [x−t,x+t] 中，
            只需要判断滑动窗口中所有大于等于 x−t 的元素中的最小元素是否小于等于 x+t 即可
            */
            auto it = slide.lower_bound((long long) nums[i] - t);
            if (it != slide.end() && abs(nums[i] - *it) <= t) {
                return true;
            }
            slide.insert(nums[i]);
            if (i >= k) {
                slide.erase(nums[i - k]);
            }
        }
        return false;
    }
};
